feat: Cluster-scoped Rollouts support

jparsai opened the pull request:
**What does this PR do / why we need it**:
This PR is to add support for cluster scoped Rollouts. Currently Rollouts controller can reconcile CR created in same namespace, this PR has changes which would allow Rollouts controller to reconcile CR created in other namespaces as well.
There are few limitations for creating Cluster scoped Rollouts controller, please check original issue created for this feature.

**Have you updated the necessary documentation?**
* [ ] Documentation update is required by this PR, and has been updated.

**Which issue(s) this PR fixes**:
Fixes #?
https://github.com/argoproj-labs/argo-rollouts-manager/issues/20

**How to test changes / Special notes to the reviewer**:
1. Create `RolloutManager` CR and change scope to namespace or cluster using `Spec.NamespaceScoped` field of `RolloutManager`.
2. Create `Rollouts` CR if `RolloutManager` is cluster scoped and it should be reconciled.
3. Try combinations of cluster and namespace scoped Rollouts controller to confirm it works as expected and doesn't allow unsupported scenarios.

jgwest commented on tests/e2e/namespace_scoped_rollouts_test.go:
```diff
@@ -0,0 +1,212 @@
+			ctx        context.Context
+			k8sClient  client.Client
+			scheme     *runtime.Scheme
+			testNsList []string
```
If I understand correctly, maybe `testNsList` -> `namespacesToDispose`?

jgwest commented on tests/e2e/namespace_scoped_rollouts_test.go:
```diff
@@ -0,0 +1,212 @@
+					return false
+				}
+				return reflect.DeepEqual(rollout.Status, rv1alpha1.RolloutStatus{})
+			}, "2m", "1s").Should(
```
Waiting `2m` is a long time (it will increase the time to complete of every test run by 2 minutes), but maybe it is necessary to wait this long? Could it be shortened, or is that not feasible, here?

jgwest commented on controllers/reconcile.go:
```diff
@@ -4,45 +4,73 @@ import (
+	log.Info("Searching for existing RolloutManager")
+	if err := checkForExistingRolloutManager(ctx, r.Client, cr, log); err != nil {
+		if doMultipleRolloutManagersExist(err) {
+			return createCondition(rolloutsmanagerv1alpha1.RolloutManagerConditionTypeErrorOccurred, metav1.ConditionTrue, rolloutsmanagerv1alpha1.RolloutManagerReasonMultipleClusterScopedRolloutManager, err.Error()), err
```
For this line, I think we can return `nil` rather than error:
`return createCondition(rolloutsmanagerv1alpha1.RolloutManagerConditionTypeErrorOccurred, metav1.ConditionTrue, rolloutsmanagerv1alpha1.RolloutManagerReasonErrorOccurred, err.Error()), nil`

We don't want to return an error in this case, because it is a user error, and thus there is nothing the we can do here. Any error that is returned will just cause the Reconcile() function to be called again and again.

(The line below this one SHOULD return nil, however, that is correct).

jgwest reviewed (changes requested):
Looks great, thanks Jayendra! See below for requested changes, but it's just minor stuff that I noticed while going the code (some of which was carried over from the original code)

In addition:
- We should add ClusterRole/Binding to `SetupWithManager`, otherwise we won't be notified about changes to these resources

- There are a few places where we are verifying that resources exist (for example, that a `Deployment` exists of name `DefaultArgoRolloutsResourceName`), but not the actual contents of those resource (for example, that the `Deployment` has the expected values inside)
	- I saw a couple spots where this was true, and it would be good to add some quick additional tests to verify resource contents:
		- In `validateArgoRolloutManagerResources` for E2E tests
		- `argorollouts_controller_test.go` (I know this came from the original test code, but can likely reuse the same code used for E2E tests here too)

jparsai commented:
Addressed all review comments, PTAL.

jgwest commented:
@jparsai did you miss pushing a commit, by chance? I started re-reviewing by checking the first [two](https://github.com/argoproj-labs/argo-rollouts-manager/pull/39#discussion_r1491687467) [comments](https://github.com/argoproj-labs/argo-rollouts-manager/pull/39#discussion_r1491691186) , and it seems like both are still present in the code.

jparsai commented:
Yes @jgwest the first comment about duplicate import I had fixed but it seems when I copied some code from another file to rollout_test.go it was added again because of auto import in VSCode.

About 2nd comment I removed the error Type and using only one type "Reconciled", but Reason has to have multiple values.
Type=Reconciled | Status =  true | Reason=Success | Message = ""
Type=Reconciled | Status =  false | Reason=ErrorOccurred | Message = "Some error message"

I removed duplicate import, PTAL.

jgwest commented:
@jparsai Perfect, it's likely I misread reason and type, thanks for double checking!

jparsai commented:
/test

jparsai commented:
@jgwest PR is ready for review now, PTAL.

jgwest commented on controllers/reconcile.go:
```diff
@@ -4,67 +4,117 @@ import (
 	log.Info("reconciling status of workloads")
 	if err := r.reconcileStatus(ctx, cr); err != nil {
-		return err
+		log.Error(err, "failed to reconcile Rollout's Metrics Service.")
```
`failed to reconcile Rollout's Metrics Service.` -> `failed to reconcile status of workloads`

jgwest reviewed (changes requested):
Looks great! All that's is just some miscellaneous small items I noticed while reviewing,

jparsai commented on controllers/utils.go:
```diff
@@ -132,3 +145,240 @@ func isMergable(extraArgs []string, cmd []string) error {
+func UpdateDeploymentStatusWithFunction(ctx context.Context, k8sClient client.Client, rm *rolloutsmanagerv1alpha1.RolloutManager,
+	mutationFn func(*rolloutsmanagerv1alpha1.RolloutManager)) error {
+
+	return untilSuccess(k8sClient, func(k8sClient client.Client) error {
+
+		// Retrieve the latest version of the RolloutManager resource
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(rm), rm)
+		if err != nil {
+			return err
+		}
+
+		// Call the mutation function, to change the RolloutManager
+		mutationFn(rm)
+
+		// Attempt to update the object with the change made by the mutation function
+		err = k8sClient.Status().Update(ctx, rm)
+
+		// Report back the error, if we hit one
+		return err
+	})
```
The issue I was talking about is that for failure in updating status condition, there will be some extra time added because of retry mechanism (it will try for given time, i.e 2 min for now, if status is not updated in this duration it will return error anyway), but I think even if we remove this and just reconcile again, then also it will take some time in running all processes again (and make all API calls and create objects again), also logs will be clean else it is filled with update errors. I personally feel retrying status update is better than reconciling all over again, if you disagree I can remove this retry mechanism and just use `.Update().Status()` function.

jparsai replied:
I am using RetryOnConflict instead of PollImmediate.
https://pkg.go.dev/k8s.io/client-go@v0.26.0/util/retry#RetryOnConflict

jgwest replied:
As written above, my preference is to follow the standard operator pattern of returning an error here, rather than using RetryOnConflict/PollImmediate. The issue with using the ROC/PI approach is there is a race condition from when a user makes a change to a resource, to when the status is set:
1) Reconcile is called on resource A, Reconcile starts performing logic
2) Reconcile retrieves a copy of resource A, version 1
2) User makes a change to resource A, producing resource A version 2
3) Reconcile from step 1 finishes (still acting on version 1), and attempts to set status on resource A
    - This fails, as resource A has changed. So, with the ROC/PI approach, Reconciles acquires the new version, and sets the status on that.
4) However, the status that Reconcile is set is only applicable to version 1, since this is the version that Reconcile acted on. 
    - We have thus (temporarily) incorrectly reported a status for an newer version of Resource A, that actually applied to an older version of Resource A.
    
In contrast, with the idiomatic Reconcile-based approach, 3 would fail, and then loop to 1 (where version 2 would be retrieved), solving the problem.

jgwest replied:
Thanks @jparsai! We'll try out this way and see if it causes us any issues, and we can definitely take a second look at our solution once we have some more data from  running and debug the product :crossed_fingers:

jparsai commented on tests/e2e/cluster-scoped/cluster_scoped_rollouts_test.go:
```diff
@@ -0,0 +1,330 @@
+			By("1st Namespace: Create and validate Rollout in 1st namespace.")
+			utils.ValidateArgoRolloutsResources(ctx, k8sClient, nsName2, 31000, 32000)
+
+			By("2nd Namespace: Create a another namespace for 2nd Rollout.")
+			Expect(utils.CreateNamespace(ctx, k8sClient, nsName3)).To(Succeed())
+
+			By("2nd Namespace: Create and validate Rollout in 2nd namespace.")
+			utils.ValidateArgoRolloutsResources(ctx, k8sClient, nsName3, 31001, 32002)
```
Actually these are two different scenarios, in 1st test case, it is checking that 2 cluster scope RMs should not be allowed and and in 2nd it is checking that namespace scoped RM should not be allowed if a cluster scoped RM is available. Both could be merged but that it will be little confusing, I think.

jparsai commented on tests/e2e/rollouts_test.go:
```diff
@@ -286,10 +212,10 @@ var _ = Describe("RolloutManager tests", func() {
+				rolloutManager.Spec.Image = controllers.DefaultArgoRolloutsImage
+				rolloutManager.Spec.Version = controllers.DefaultArgoRolloutsVersion
+
 				Expect(k8sClient.Update(ctx, &rolloutManager)).To(Succeed())
```
It will use the same function as this https://github.com/argoproj-labs/argo-rollouts-manager/pull/39#discussion_r1550776597. Should we remove retry mechanism from controller but still use it for tests?

jgwest replied:
Yah, my suggestion is to remove the retry mechanism from the controller, but keep it for the tests:
- The Reconcile() loop is a standard operator pattern, where, if an error occurs, the resource will be re-reconciled and the logic will run again. This necessarily means *re-running* some of the logic and API calls again, but this is very little work: API calls are just simple GET requests from the operator to etcd, and the logic that runs is trivial.
    - Ultimately the goal is, with each Reconcile run, the resource gets closer to it's desired, declaratively defined state.
- In contrast, for the tests, we don't have an equivalent to Reconcile mechanism, so we need to use our own retry loop.

jgwest commented on tests/e2e/namespa-scoped/namespa_scoped_suite_test.go:
```diff
@@ -0,0 +1,23 @@
+	logf.SetLogger(zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true), zap.Level(zapcore.DebugLevel)))
+})
+
+func TestNamespaScoped(t *testing.T) {
```
Looks like the suite name `TestNamespaScoepd` still needs to be updated
